auth/realm/base/src/main/java/org/wildfly/security/auth/realm/JaasSecurityRealm.java

```diff
@@ -211,80 +257,160 @@ public boolean exists() throws RealmUnavailableException {
+         * on November 2021
+         * <p>
+         * Try to convert the credential value into a char[] using the
+         * first of the following attempts which succeeds:
+         * <p>
+         * 1. Check for instanceof char[]
+         * 2. Check for instanceof String and then use toCharArray()
+         * 3. See if credential has a toCharArray() method and use it
+         * 4. Use toString() followed by toCharArray().
+         *
+         * @return a char[] representation of the credential.
+         */
+        private char[] getPassword() {
+            char[] password = null;
+            if (evidence instanceof char[]) {
+                password = (char[]) evidence;
+            } else if (evidence instanceof String) {
+                String s = (String) evidence;
+                password = s.toCharArray();
+            } else {
+                try {
+                    Class<?>[] types = {};
+                    Method m = evidence.getClass().getMethod("toCharArray", types);
+                    Object[] args = {};
+                    password = (char[]) m.invoke(evidence, args);
+                } catch (Exception e) {
+                    if (evidence != null) {
+                        String s = evidence.toString();
+                        password = s.toCharArray();
+                    }
                 }
             }
+            return password;
         }
     }
 
+    /**
+     * A JAAS realm's authorization identity. Roles are mapped from all Subject's principals with the following rule:
+     * key of the attribute is principal's simple classname and the value is principal's name
+     */
     private static class JaasAuthorizationIdentity implements AuthorizationIdentity {
 
-        private static final String CALLER_PRINCIPAL_GROUP = "CallerPrincipal";
+        private final MapAttributes attributes;
 
-        private final Principal principal;
-        private Principal callerPrincipal;
-        private final Subject subject;
+        private static JaasAuthorizationIdentity fromSubject(final Subject subject) {
 
-        private JaasAuthorizationIdentity(final Principal principal, final Subject subject) {
-            this.principal = principal;
-            this.subject = subject;
-            // check if the subject has a caller principal group - if it has then we should use that principal.
-            this.callerPrincipal = getCallerPrincipal(subject);
+            MapAttributes attributes = new MapAttributes();
+            // map all subject's principals to attributes with the following rule:
+            // key of the attribute is principal's simple classname and the value is principal's name
+            for (Principal principal : subject.getPrincipals()) {
+                attributes.addLast(principal.getClass().getSimpleName(), principal.getName());
```

Comment:
@darranl Just FYI, this is how I am currenly mapping subject's principal list into attributes of SecurityIdentity. It is like you mentioned on the meeting - attribute key is **simple classname** of Principal and attribute value is **principal's name**.